lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -6,7 +6,182 @@
+
+      getter = lambda do |provider|
+        munge.call(provider.user[property])
+      end
+      setter = lambda do |provider, value|
+        provider.user[property] = unmunge.call(value)
+      end
+
+      info = info.merge(getter: getter, setter: setter)
+
+      attribute(info)
+    end
+
+    # Adds a new attribute that directly maps to an ADS User flag. The info
+    # hash must include:
+    #
+    #   * :name -- The attribute's name
+    #   * :flag -- The ADS User flag that the attribute maps to
+    #
+    def userflag_attribute(info = {})
+      flag = info.delete(:flag)
+
+      getter = lambda do |provider|
+        provider.user.userflag_set?(flag).to_s
+      end
+      setter = lambda do |provider, set_flag|
+        unless set_flag == 'true' || set_flag == 'false'
+          raise ArgumentError, _("'%{set_flag}' is not a Boolean value! Boolean values are 'true' or 'false'") % { set_flag: set_flag }
+        end
+
+        if set_flag == 'true'
+          provider.user.set_userflags(flag)
+        else
+          provider.user.unset_userflags(flag)
+        end
+      end
+
+      info = info.merge(
+        getter: getter,
+        setter: setter
+      )
+
+      attribute(info)
+    end
+
+    def munge_string(value)
+      value
+    end
+
+    def unmunge_string(value)
+      # WIN32OLE will throw an unreadable error message if we try
+      # to set the attribute value to the empty string, so we check
+      # for this case and throw a more readable one here.
+      if value.empty?
+        raise ArgumentError, "The empty string is not an allowable attribute value!"
+      end
+
+      value
+    end
+
+    def munge_bit(value)
+      return 'true' if value == 1
+      return 'false' if value == 0
+      raise ArgumentError, _("'%{value}' is not a bit value! Bit values are 0 or 1") % { value: value }
+    end
+
+    def unmunge_bit(value)
+      return 1 if value == 'true'
+      return 0 if value == 'false'
+      raise ArgumentError, _("'%{value}' is not a Boolean value! Boolean values are 'true' or 'false'") % { value: value }
+    end
+  end
+
+  # We can also manage many more attributes than what's listed here. All of
+  # the userflags can probably be individually managed. See
+  #   https://docs.microsoft.com/en-us/windows/desktop/api/iads/ne-iads-ads_user_flag
+  #
+  # We can also manage the AccountExpirationDate, and possibly some of the
+  # properties in https://docs.microsoft.com/en-us/windows/desktop/ADSI/winnt-custom-user-properties
+
+  adsi_property_attribute name: :full_name,
+                          property: 'FullName'
+
+  adsi_property_attribute name: :password_change_required,
+                          property: 'PasswordExpired',
+                          munge: method(:munge_bit),
+                          unmunge: method(:unmunge_bit)
+
+  userflag_attribute name: :disabled,
+                     flag: :ADS_UF_ACCOUNTDISABLE
+
+  userflag_attribute name: :password_change_not_allowed,
```

Comment:
What's the rationale behind the name `password_change_not_allowed`? It seems like something positive instead of negative (like `password_change_allowed`) would be easier to reason about.

Reply:
I guess that matches the ADS flag - if it's hard to change, no problem, I guess

Reply:
It also matches the corresponding DSC User resource attribute -- https://docs.microsoft.com/en-us/powershell/dsc/userresource (that's where the attribute names come from)